diff --git a/net/transmission/files/transmission.init b/net/transmission/files/transmission.init
index 521fe20..1f54a4e 100644
--- a/net/transmission/files/transmission.init
+++ b/net/transmission/files/transmission.init
@@ -140,3 +140,6 @@ start_service() {
 	config_foreach transmission 'transmission'
 }
 
+service_triggers() {
+	procd_add_reload_trigger 'transmission'
+}
diff --git a/libs/tcp_wrappers/Makefile b/libs/tcp_wrappers/Makefile
index ca9bfd2..7a1e447 100644
--- a/libs/tcp_wrappers/Makefile
+++ b/libs/tcp_wrappers/Makefile
@@ -32,7 +32,7 @@ endef
 
 TARGET_CFLAGS += $(FPIC)
 
-ifeq ($(CONFIG_USE_MUSL),)
+ifneq ($(CONFIG_USE_UCLIBC),)
 TARGET_EXTRA_LIBS:=LIBS=-lnsl
 endif
 
