diff --git a/net/samba4/Config.in b/net/samba4/Config.in
index 17a6cf2..10cbdc3 100644
--- a/net/samba4/Config.in
+++ b/net/samba4/Config.in
@@ -57,3 +57,10 @@ config SAMBA4_SERVER_QUOTAS
 
 		Support for disk quotas using the quotas VFS module (vfs_default_quota)
 	default n
+
+config SAMBA4_SERVER_HOTPLUG
+	bool "enable HOTPLUG support"
+	depends on PACKAGE_samba4-server
+	help
+		Provide hotplug support.
+	default y
diff --git a/net/samba4/Makefile b/net/samba4/Makefile
index b4e7690..953f37e 100644
--- a/net/samba4/Makefile
+++ b/net/samba4/Makefile
@@ -30,6 +30,7 @@ PKG_CONFIG_DEPENDS:= \
 	CONFIG_SAMBA4_SERVER_VFSX \
 	CONFIG_SAMBA4_SERVER_QUOTAS \
 	CONFIG_SAMBA4_SERVER_ACL \
+	CONFIG_SAMBA4_SERVER_HOTPLUG \
 	CONFIG_PACKAGE_kmod-fs-btrfs \
 	CONFIG_PACKAGE_kmod-fs-xfs
 
@@ -165,7 +166,6 @@ CONFIGURE_ARGS += \
 		--disable-rpath-private-install \
 		--enable-fhs \
 		--without-automount \
-		--without-iconv \
 		--without-lttng \
 		--without-ntvfs-fileserver \
 		--without-pam \
@@ -397,6 +397,12 @@ endif
 	$(INSTALL_CONF) ./files/samba.config $(1)/etc/config/samba4
 	$(INSTALL_DATA) ./files/smb.conf.template $(1)/etc/samba
 	$(INSTALL_BIN) ./files/samba.init $(1)/etc/init.d/samba4
+ifeq ($(CONFIG_SAMBA4_SERVER_HOTPLUG),y)
+	$(INSTALL_DIR) $(1)/lib/samba
+	$(INSTALL_DATA) ./files/samba.sh $(1)/lib/samba/samba.sh
+	$(INSTALL_DIR) $(1)/etc/hotplug.d/block
+	$(INSTALL_DATA) ./files/samba.hotplug $(1)/etc/hotplug.d/block/60-samba
+endif
 endef
 
 define Package/samba4-server/conffiles
diff --git a/net/samba4/files/samba.hotplug b/net/samba4/files/samba.hotplug
new file mode 100644
index 0000000..111f0cd
--- /dev/null
+++ b/net/samba4/files/samba.hotplug
@@ -0,0 +1,11 @@
+. /lib/samba/samba.sh
+
+samba_dev_filter $DEVNAME
+[ "$FLAG_DEV_TYPE" = "not-mtd" ]  || exit
+
+config_load samba4
+case $ACTION in
+	add) samba_add_section $DEVNAME;;
+	remove) samba_delete_section $DEVNAME;;
+esac
+uci commit samba4
diff --git a/net/samba4/files/samba.sh b/net/samba4/files/samba.sh
new file mode 100644
index 0000000..5416cbc
--- /dev/null
+++ b/net/samba4/files/samba.sh
@@ -0,0 +1,89 @@
+#!/bin/sh
+#
+# Copyright (C) 2018 OpenWrt.org
+# Copyright (C) 2018 rosysong@rosinson.com
+#
+
+. /lib/functions.sh
+
+FLAG_DEV_TYPE=
+FLAG_DEV_MOPT=
+FLAG_HAS_SECT=
+
+samba_dev_filter() { # <devname> <[path,/dev/]>
+	case $1 in
+	${2}mtdblock*|\
+	${2}ubi*)
+		FLAG_DEV_TYPE="mtd"
+		;;
+	${2}loop*|\
+	${2}mmcblk*|\
+	${2}sd*|\
+	${2}hd*|\
+	${2}md*|\
+	${2}nvme*|\
+	${2}vd*|\
+	${2}xvd*)
+		FLAG_DEV_TYPE="not-mtd"
+		;;
+	*)
+		[ -b ${2}${1} ] && FLAG_DEV_TYPE="not-mtd"
+		[ -b /dev/mapper/$1 ] && FLAG_DEV_TYPE="not-mtd"
+		;;
+	esac
+}
+
+samba_cfg_lookup() { # <section> <name>
+	config_get name $1 name
+	[ "$name" = "$2" ] || return
+	FLAG_HAS_SECT=y
+}
+
+samba_cfg_delete() { # <section> <name>
+	config_get name $1 name
+	[ "$name" = "$2" ] || return
+	uci -q delete samba4.$1
+}
+
+samba_find_mount_point() { # <devname>
+	# search mount point in /proc/mounts
+	while read l; do
+		local d=$(echo $l | awk '/^\/dev/ {print $1}')
+		[ "$d" = "/dev/$1" ] || continue
+
+		FLAG_DEV_MOPT=$(echo $l | awk '/^\/dev/ {print $2}')
+		break
+	done < /proc/mounts
+}
+
+_samba_add_section() { # <devname> <mount point>
+	uci -q batch <<-EOF
+		add samba4 sambashare
+		set samba4.@sambashare[-1].browseable='yes'
+		set samba4.@sambashare[-1].name='$1'
+		set samba4.@sambashare[-1].path='$2'
+		set samba4.@sambashare[-1].users='root'
+		set samba4.@sambashare[-1].read_only='no'
+		set samba4.@sambashare[-1].guest_ok='yes'
+		set samba4.@sambashare[-1].create_mask='0755'
+		set samba4.@sambashare[-1].dir_mask='0755'
+	EOF
+}
+
+samba_add_section() { # <devname> [<mount point>]
+	FLAG_HAS_SECT=
+	FLAG_DEV_MOPT=
+
+	config_foreach samba_cfg_lookup sambashare $1
+	[ -z "$FLAG_HAS_SECT" ] || return
+
+	samba_find_mount_point $1
+	[ -n "$FLAG_DEV_MOPT" ] || return
+
+	[ -n "$2" -a "$2" = "$FLAG_DEV_MOPT" ] || \
+		_samba_add_section $1 $FLAG_DEV_MOPT
+}
+
+samba_delete_section() { # <devname>
+	config_foreach samba_cfg_delete sambashare $1
+}
diff --git a/net/samba4/samba4-libs/etc/hotplug.d/block/60-samba b/net/samba4/samba4-libs/etc/hotplug.d/block/60-samba
new file mode 100644
index 0000000..111f0cd
--- /dev/null
+++ b/net/samba4/samba4-libs/etc/hotplug.d/block/60-samba
@@ -0,0 +1,11 @@
+. /lib/samba/samba.sh
+
+samba_dev_filter $DEVNAME
+[ "$FLAG_DEV_TYPE" = "not-mtd" ]  || exit
+
+config_load samba4
+case $ACTION in
+	add) samba_add_section $DEVNAME;;
+	remove) samba_delete_section $DEVNAME;;
+esac
+uci commit samba4
diff --git a/net/samba4/samba4-libs/lib/samba/samba.sh b/net/samba4/samba4-libs/lib/samba/samba.sh
new file mode 100644
index 0000000..5416cbc
--- /dev/null
+++ b/net/samba4/samba4-libs/lib/samba/samba.sh
@@ -0,0 +1,89 @@
+#!/bin/sh
+#
+# Copyright (C) 2018 OpenWrt.org
+# Copyright (C) 2018 rosysong@rosinson.com
+#
+
+. /lib/functions.sh
+
+FLAG_DEV_TYPE=
+FLAG_DEV_MOPT=
+FLAG_HAS_SECT=
+
+samba_dev_filter() { # <devname> <[path,/dev/]>
+	case $1 in
+	${2}mtdblock*|\
+	${2}ubi*)
+		FLAG_DEV_TYPE="mtd"
+		;;
+	${2}loop*|\
+	${2}mmcblk*|\
+	${2}sd*|\
+	${2}hd*|\
+	${2}md*|\
+	${2}nvme*|\
+	${2}vd*|\
+	${2}xvd*)
+		FLAG_DEV_TYPE="not-mtd"
+		;;
+	*)
+		[ -b ${2}${1} ] && FLAG_DEV_TYPE="not-mtd"
+		[ -b /dev/mapper/$1 ] && FLAG_DEV_TYPE="not-mtd"
+		;;
+	esac
+}
+
+samba_cfg_lookup() { # <section> <name>
+	config_get name $1 name
+	[ "$name" = "$2" ] || return
+	FLAG_HAS_SECT=y
+}
+
+samba_cfg_delete() { # <section> <name>
+	config_get name $1 name
+	[ "$name" = "$2" ] || return
+	uci -q delete samba4.$1
+}
+
+samba_find_mount_point() { # <devname>
+	# search mount point in /proc/mounts
+	while read l; do
+		local d=$(echo $l | awk '/^\/dev/ {print $1}')
+		[ "$d" = "/dev/$1" ] || continue
+
+		FLAG_DEV_MOPT=$(echo $l | awk '/^\/dev/ {print $2}')
+		break
+	done < /proc/mounts
+}
+
+_samba_add_section() { # <devname> <mount point>
+	uci -q batch <<-EOF
+		add samba4 sambashare
+		set samba4.@sambashare[-1].browseable='yes'
+		set samba4.@sambashare[-1].name='$1'
+		set samba4.@sambashare[-1].path='$2'
+		set samba4.@sambashare[-1].users='root'
+		set samba4.@sambashare[-1].read_only='no'
+		set samba4.@sambashare[-1].guest_ok='yes'
+		set samba4.@sambashare[-1].create_mask='0755'
+		set samba4.@sambashare[-1].dir_mask='0755'
+	EOF
+}
+
+samba_add_section() { # <devname> [<mount point>]
+	FLAG_HAS_SECT=
+	FLAG_DEV_MOPT=
+
+	config_foreach samba_cfg_lookup sambashare $1
+	[ -z "$FLAG_HAS_SECT" ] || return
+
+	samba_find_mount_point $1
+	[ -n "$FLAG_DEV_MOPT" ] || return
+
+	[ -n "$2" -a "$2" = "$FLAG_DEV_MOPT" ] || \
+		_samba_add_section $1 $FLAG_DEV_MOPT
+}
+
+samba_delete_section() { # <devname>
+	config_foreach samba_cfg_delete sambashare $1
+}
diff --git a/net/samba4/waf-cross-answers/arm.txt b/net/samba4/waf-cross-answers/arm.txt
index 79ba994..728795f 100644
--- a/net/samba4/waf-cross-answers/arm.txt
+++ b/net/samba4/waf-cross-answers/arm.txt
@@ -1,11 +1,15 @@
 Checking uname sysname type: "Linux"
+Checking uname machine type: "armv7l"
+Checking uname release type: "4.15.0-46-generic"
+Checking uname version type: "#49~16.04.1-Ubuntu SMP Tue Feb 12 17:45:24 UTC 2019"
 Checking simple C program: "hello world"
 rpath library support: OK
 -Wl,--version-script support: OK
 Checking getconf LFS_CFLAGS: NO
-Checking for large file support without additional flags: OK
-Checking correct behavior of strtoll: OK
-Checking for working strptime: NO
+Checking for large file support without additional flags: NO
+Checking for -D_FILE_OFFSET_BITS=64: OK
+Checking correct behavior of strtoll: NO
+Checking for working strptime: OK
 Checking for C99 vsnprintf: "1"
 Checking for HAVE_SHARED_MMAP: OK
 Checking for HAVE_MREMAP: OK
@@ -15,17 +19,18 @@ Checking for HAVE_IFACE_GETIFADDRS: OK
 Checking value of NSIG: "65"
 Checking value of _NSIG: "65"
 Checking value of SIGRTMAX: "64"
-Checking value of SIGRTMIN: "35"
+Checking value of SIGRTMIN: "34"
 Checking for a 64-bit host to support lmdb: NO
-Checking whether the WRFILE -keytab is supported: OK
-Checking errno of iconv for illegal multibyte sequence: OK
+Checking errno of iconv for illegal multibyte sequence: (255, "")
 Checking for kernel change notify support: OK
 Checking for Linux kernel oplocks: OK
 Checking for kernel share modes: OK
-Checking if can we convert from CP850 to UCS-2LE: OK
+Checking whether POSIX capabilities are available: OK
+Checking if can we convert from CP850 to UCS-2LE: (255, "")
+Checking if can we convert from IBM850 to UCS-2LE: (255, "")
 Checking if can we convert from UTF-8 to UCS-2LE: OK
 vfs_fileid checking for statfs() and struct statfs.f_fsid: OK
-Checking whether we can use Linux thread-specific credentials with 32-bit system calls: "OK"
+Checking whether we can use Linux thread-specific credentials with 32-bit system calls: OK
 Checking whether fcntl locking is available: OK
 Checking whether fcntl lock supports open file description locks: NO
 Checking for the maximum value of the 'time_t' type: NO
